File: backend/app/Models/Transaction.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Transaction extends Model
{
    use HasFactory;

    /**
     * The attributes that are mass assignable.
     *
     * @var array<int, string>
     */
    protected $fillable = [
        'category_id',
        'amount',
        'transaction_date'
    ];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'amount' => 'decimal:2',
        'transaction_date' => 'date',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    /**
     * Get the category that owns the transaction.
     */
    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class);
    }

    /**
     * Scope a query to only include expenses
     */
    public function scopeExpenses($query)
    {
        return $query->whereHas('category', function ($q) {
            $q->where('type', 'expense');
        });
    }

    /**
     * Scope a query to only include income
     */
    public function scopeIncome($query)
    {
        return $query->whereHas('category', function ($q) {
            $q->where('type', 'income');
        });
    }

    /**
     * Scope a query to filter by date range
     */
    public function scopeDateBetween($query, $startDate, $endDate)
    {
        return $query->whereBetween('transaction_date', [$startDate, $endDate]);
    }

    /**
     * Get transaction type through category
     */
    public function getTypeAttribute()
    {
        return $this->category->type;
    }

    /**
     * Format amount with currency
     */
    public function getFormattedAmountAttribute()
    {
        return 'Rp ' . number_format($this->amount, 2, ',', '.');
    }
}

File: frontend_app/lib/models/transaction.dart
import '../models/category.dart';

class Transaction {
  int? id;
  int categoryId;
  double amount;
  String transactionDate;

  Category? category;

  Transaction({
    this.id,
    required this.categoryId,
    required this.amount,
    required this.transactionDate,
    this.category,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'],
      categoryId: json['category_id'],
      amount: double.parse(json['amount'].toString()),
      transactionDate: json['transaction_date'],
      category:
          json['category'] != null ? Category.fromJson(json['category']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'category_id': categoryId,
      'amount': amount,
      'transaction_date': transactionDate
    };
  }
}
